runtime/Cpp/runtime/src/atn/SingletonPredictionContext.cpp

```diff
@@ -45,11 +49,11 @@ bool SingletonPredictionContext::operator == (const PredictionContext &o) const
     return true;
   }
 
-  const SingletonPredictionContext *other = dynamic_cast<const SingletonPredictionContext*>(&o);
-  if (other == nullptr) {
+  if (o.getContextType() != PredictionContextType::EMPTY && o.getContextType() != PredictionContextType::SINGLETON) {
```

Comment:
Why do we need the extra test for ::EMPTY here?

Reply:
Fixed, we do not now. I removed EmptyPredictionContext as its not useful and can be implemented purely in terms of SIngletonPredictionContext. It also avoids people instantiating EmptyPredictionContext, since some code uses pointer equality to the PredictionContext::EMPTY.